File: plugin/engine/Music.cpp
#include "Music.h"

#include "Num.h"

#include <array>
#include <cstdlib>

namespace idm
{

namespace
{

constexpr int chromaticDegrees[]      = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11 };
constexpr int majorDegrees[]          = { 0, 2, 4, 5, 7, 9, 11 };
constexpr int minorDegrees[]          = { 0, 2, 3, 5, 7, 8, 10 };
constexpr int dorianDegrees[]         = { 0, 2, 3, 5, 7, 9, 10 };
constexpr int mixolydianDegrees[]     = { 0, 2, 4, 5, 7, 9, 10 };
constexpr int lydianDegrees[]         = { 0, 2, 4, 6, 7, 9, 11 };
constexpr int phrygianDegrees[]       = { 0, 1, 3, 5, 7, 8, 10 };
constexpr int harmonicMinorDegrees[]  = { 0, 2, 3, 5, 7, 8, 11 };
constexpr int majorPentatonicDegrees[]= { 0, 2, 4, 7, 9 };
constexpr int minorPentatonicDegrees[]= { 0, 3, 5, 7, 10 };
constexpr int bluesDegrees[]          = { 0, 3, 5, 6, 7, 10 };

constexpr const char* noteNames[] = {
    "C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"
};

/** The signed pitch-class distance from `pc` to `targetPc`, folded into
    [-6, 5] so ties resolve downward. */
int signedDelta (int targetPc, int pc) noexcept
{
    return floorMod (targetPc - pc + 6, 12) - 6;
}

/** The degree nearest `rel`, measured around the twelve-tone circle. */
int nearestDegree (const int* degrees, int count, int rel) noexcept
{
    auto best = degrees[0];
    auto bestDist = 99;

    for (int i = 0; i < count; ++i)
    {
        const auto raw = std::abs (rel - degrees[i]);
        const auto dist = raw < 12 - raw ? raw : 12 - raw;

        if (dist < bestDist)
        {
            bestDist = dist;
            best = degrees[i];
        }
    }

    return best;
}

} // namespace

int floorMod (int value, int modulus) noexcept
{
    const auto r = value % modulus;
    return r < 0 ? r + modulus : r;
}

int floorDiv (int value, int divisor) noexcept
{
    auto q = value / divisor;

    if ((value % divisor != 0) && ((value < 0) != (divisor < 0)))
        --q;

    return q;
}

Degrees degreesOf (Scale scale) noexcept
{
    switch (scale)
    {
        case Scale::chromatic:       return { chromaticDegrees, 12 };
        case Scale::major:           return { majorDegrees, 7 };
        case Scale::minor:           return { minorDegrees, 7 };
        case Scale::dorian:          return { dorianDegrees, 7 };
        case Scale::mixolydian:      return { mixolydianDegrees, 7 };
        case Scale::lydian:          return { lydianDegrees, 7 };
        case Scale::phrygian:        return { phrygianDegrees, 7 };
        case Scale::harmonicMinor:   return { harmonicMinorDegrees, 7 };
        case Scale::majorPentatonic: return { majorPentatonicDegrees, 5 };
        case Scale::minorPentatonic: return { minorPentatonicDegrees, 5 };
        case Scale::blues:           return { bluesDegrees, 6 };
    }

    return { chromaticDegrees, 12 };
}

const char* nameOf (Scale scale) noexcept
{
    switch (scale)
    {
        case Scale::chromatic:       return "chromatic";
        case Scale::major:           return "major";
        case Scale::minor:           return "minor";
        case Scale::dorian:          return "dorian";
        case Scale::mixolydian:      return "mixolydian";
        case Scale::lydian:          return "lydian";
        case Scale::phrygian:        return "phrygian";
        case Scale::harmonicMinor:   return "harmonicMinor";
        case Scale::majorPentatonic: return "majorPentatonic";
        case Scale::minorPentatonic: return "minorPentatonic";
        case Scale::blues:           return "blues";
    }

    return "chromatic";
}

bool scaleFromName (const std::string& name, Scale& out) noexcept
{
    static const std::array<Scale, 11> all {
        Scale::chromatic, Scale::major, Scale::minor, Scale::dorian,
        Scale::mixolydian, Scale::lydian, Scale::phrygian, Scale::harmonicMinor,
        Scale::majorPentatonic, Scale::minorPentatonic, Scale::blues
    };

    for (auto scale : all)
    {
        if (name == nameOf (scale))
        {
            out = scale;
            return true;
        }
    }

    return false;
}

std::string midiToName (int note)
{
    const auto octave = floorDiv (note, 12) - 1;
    return std::string (noteNames[floorMod (note, 12)]) + std::to_string (octave);
}

int snapToScale (int note, int root, Scale scale) noexcept
{
    const auto degrees = degreesOf (scale);

    if (degrees.count == 12)
        return note; // chromatic: no change

    const auto pc = floorMod (note, 12);
    const auto rootPc = floorMod (root, 12);
    const auto rel = floorMod (pc - rootPc, 12);

    const auto best = nearestDegree (degrees.values, degrees.count, rel);
    return note + signedDelta ((rootPc + best) % 12, pc);
}

int snapToChord (int note, int root, Scale scale) noexcept
{
    const auto degrees = degreesOf (scale);

    int chord[3];

    if (degrees.count == 12)
    {
        chord[0] = 0; chord[1] = 4; chord[2] = 7;
    }
    else
    {
        chord[0] = degrees[0]; chord[1] = degrees[2]; chord[2] = degrees[4];
    }

    const auto pc = floorMod (note, 12);
    const auto rootPc = floorMod (root, 12);
    const auto rel = floorMod (pc - rootPc, 12);

    const auto best = nearestDegree (chord, 3, rel);
    return note + signedDelta ((rootPc + best) % 12, pc);
}

int diatonicTranspose (int note, int root, Scale scale, int steps) noexcept
{
    const auto degrees = degreesOf (scale);

    if (degrees.count == 12)
        return note + steps;

    const auto rootPc = floorMod (root, 12);
    const auto snapped = snapToScale (note, root, scale);
    const auto relPc = floorMod (snapped - rootPc, 12);

    // `indexOf` yields -1 when absent. Mirrored rather than tidied away: the
    // arithmetic below is defined for it and the TypeScript relies on that.
    auto di = -1;

    for (int i = 0; i < degrees.count; ++i)
        if (degrees[i] == relPc)
        {
            di = i;
            break;
        }

    const auto len = degrees.count;
    const auto total = di + steps;
    const auto octave = floorDiv (total, len);
    const auto wrapped = floorMod (total, len);

    return snapped - relPc + degrees[wrapped] + 12 * octave;
}

int clampMidi (double note) noexcept
{
    return (int) clampTo (jsRound (note), 0.0, 127.0);
}

} // namespace idm

File: engine-rust-parked/src/music.rs
//! Ported from `src/engine/music.ts`.
//!
//! Scales, snapping, and diatonic transposition. All of the arithmetic here is
//! pitch-class modular, and JavaScript's `%` keeps the sign of the dividend —
//! so every wrap uses `rem_euclid` rather than `%`, and every `Math.floor` of a
//! quotient uses `div_euclid`. A negative note is where a careless port breaks.

use crate::num::js_round;

pub const NOTE_NAMES: [&str; 12] = [
    "C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B",
];

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum Scale {
    Chromatic,
    Major,
    Minor,
    Dorian,
    Mixolydian,
    Lydian,
    Phrygian,
    HarmonicMinor,
    MajorPentatonic,
    MinorPentatonic,
    Blues,
}

impl Scale {
    /// Scale degrees as semitone offsets from the root.
    pub fn degrees(self) -> &'static [i32] {
        match self {
            Scale::Chromatic => &[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
            Scale::Major => &[0, 2, 4, 5, 7, 9, 11],
            Scale::Minor => &[0, 2, 3, 5, 7, 8, 10],
            Scale::Dorian => &[0, 2, 3, 5, 7, 9, 10],
            Scale::Mixolydian => &[0, 2, 4, 5, 7, 9, 10],
            Scale::Lydian => &[0, 2, 4, 6, 7, 9, 11],
            Scale::Phrygian => &[0, 1, 3, 5, 7, 8, 10],
            Scale::HarmonicMinor => &[0, 2, 3, 5, 7, 8, 11],
            Scale::MajorPentatonic => &[0, 2, 4, 7, 9],
            Scale::MinorPentatonic => &[0, 3, 5, 7, 10],
            Scale::Blues => &[0, 3, 5, 6, 7, 10],
        }
    }

    pub fn from_name(name: &str) -> Option<Self> {
        Some(match name {
            "chromatic" => Scale::Chromatic,
            "major" => Scale::Major,
            "minor" => Scale::Minor,
            "dorian" => Scale::Dorian,
            "mixolydian" => Scale::Mixolydian,
            "lydian" => Scale::Lydian,
            "phrygian" => Scale::Phrygian,
            "harmonicMinor" => Scale::HarmonicMinor,
            "majorPentatonic" => Scale::MajorPentatonic,
            "minorPentatonic" => Scale::MinorPentatonic,
            "blues" => Scale::Blues,
            _ => return None,
        })
    }
}

pub fn midi_to_name(note: i32) -> String {
    let octave = note.div_euclid(12) - 1;
    format!("{}{}", NOTE_NAMES[note.rem_euclid(12) as usize], octave)
}

/// The signed pitch-class distance from `pc` to `target_pc`, folded into
/// [-6, 5] so ties resolve downward.
fn signed_delta(target_pc: i32, pc: i32) -> i32 {
    (target_pc - pc + 6).rem_euclid(12) - 6
}

/// The degree nearest `rel`, measured around the twelve-tone circle.
fn nearest_degree(degrees: &[i32], rel: i32) -> i32 {
    let mut best = degrees[0];
    let mut best_dist = 99;

    for &d in degrees {
        let dist = (rel - d).abs().min(12 - (rel - d).abs());
        if dist < best_dist {
            best_dist = dist;
            best = d;
        }
    }

    best
}

/// Snap a MIDI note into the given key and scale. Ties resolve downward.
pub fn snap_to_scale(note: i32, root: i32, scale: Scale) -> i32 {
    let degrees = scale.degrees();
    if degrees.len() == 12 {
        return note; // chromatic: no change
    }

    let pc = note.rem_euclid(12);
    let root_pc = root.rem_euclid(12);
    let rel = (pc - root_pc).rem_euclid(12);

    let best = nearest_degree(degrees, rel);
    note + signed_delta((root_pc + best) % 12, pc)
}

/// Snap to the nearest tone of the key's tonic triad.
pub fn snap_to_chord(note: i32, root: i32, scale: Scale) -> i32 {
    let degrees = scale.degrees();
    let chord: [i32; 3] = if degrees.len() == 12 {
        [0, 4, 7]
    } else {
        [degrees[0], degrees[2], degrees[4]]
    };

    let pc = note.rem_euclid(12);
    let root_pc = root.rem_euclid(12);
    let rel = (pc - root_pc).rem_euclid(12);

    let best = nearest_degree(&chord, rel);
    note + signed_delta((root_pc + best) % 12, pc)
}

/// Move `steps` scale degrees, folding through the key. In chromatic, steps are
/// plain semitones.
pub fn diatonic_transpose(note: i32, root: i32, scale: Scale, steps: i32) -> i32 {
    let degrees = scale.degrees();
    if degrees.len() == 12 {
        return note + steps;
    }

    let root_pc = root.rem_euclid(12);
    let snapped = snap_to_scale(note, root, scale);
    let rel_pc = (snapped - root_pc).rem_euclid(12);

    // `indexOf` yields -1 when absent; mirrored rather than tidied away, since
    // the arithmetic below is defined for it and the TypeScript relies on that.
    let di = degrees.iter().position(|&d| d == rel_pc).map_or(-1, |i| i as i32);

    let len = degrees.len() as i32;
    let total = di + steps;
    let octave = total.div_euclid(len);
    let wrapped = total.rem_euclid(len);

    snapped - rel_pc + degrees[wrapped as usize] + 12 * octave
}

pub fn clamp_midi(note: f64) -> i32 {
    js_round(note).clamp(0.0, 127.0) as i32
}
